custom/aaa/models.py

```diff
@@ -0,0 +1,272 @@
+
+
+class LocationDenormalizedModel(models.Model):
+    """Abstract base model for containing all the location fields necessary for querying."""
+
+    # support multiple domains so QA is easier
+    domain = models.TextField(null=False)
+
+    # Shared location hierarchy
+    state_id = models.TextField(null=True)
+    district_id = models.TextField(null=True)
+
+    # CAS Hierarchy
+    block_id = models.TextField(null=True)
+    supervisor_id = models.TextField(null=True)
+    awc_id = models.TextField(null=True)
+
+    # AAA/MoH Hierarchy
+    taluka_id = models.TextField(null=True)
+    phc_id = models.TextField(null=True)
+    sc_id = models.TextField(null=True)
+    village_id = models.TextField(null=True)
+
+    class Meta(object):
+        abstract = True
+
+
+class Woman(LocationDenormalizedModel):
+    """Represents a woman registered in the AAA Convergence program.
+
+    This woman may have an eligible_couple or ccs_record case associated with her.
+    ccs_record data will be attached to the CcsRecord model.
+    """
+
+    household_case_id = models.TextField(null=True)
+    person_case_id = models.TextField(primary_key=True)
+
+    opened_on = models.DateField()
+    closed_on = models.DateField(null=True)
+    pregnant_ranges = ArrayField(
+        DateRangeField(),
+        help_text="The ranges in which a ccs_record has been opened and the baby has not been born",
+        null=True
+    )
+    dob = models.DateField(null=True)
+    marital_status = models.TextField(null=True)
+    sex = models.TextField(null=True)
+    migration_status = models.TextField(null=True)
+
+    fp_current_method_ranges = ArrayField(
+        DateRangeField(),
+        help_text="Ranges of time when eligible_couple.fp_current_method != 'none'",
+        null=True
+    )
+
+
+class WomanHistory(models.Model):
+    """The history of form properties for any woman registered."""
+
+    person_case_id = models.TextField(primary_key=True)
+
+    # eligible_couple properties
+    fp_current_method_history = ArrayField(ArrayField(models.TextField(), size=2), null=True)
+    fp_preferred_method_history = ArrayField(ArrayField(models.TextField(), size=2), null=True)
+
+
+class CcsRecord(LocationDenormalizedModel):
+    """Represent a single pregnancy, lactation, and complementary feeding schedule for a Woman.
+
+    Lactation is tracked 6 months after the delivery of the child.
+    Complementary feeding is tracked until the child is 2 years old
+
+    A Woman will likely have many pregnancies over her lifetime.
+    If a Woman becomes pregnant within two years of a child being born, she will have two ccs_record cases.
+    If a Woman has twins and those twins are registered in the program before birth,
+        she will have one ccs_record for this pregnancy.
+    If a Woman has twins and those twins are not registered in the program before birth,
+        she will have two ccs_records for this pregnancy
+    """
+
+    person_case_id = models.TextField()
```

Comment:
should this be a foreign key?

Reply:
I've avoided foreign keys for now as I'm not clear on all the implications with Citus and our workflows there. I wanted to make this with a migration in mind and wanted to keep it simple for that.